### DISCUSSION TITLE
Tuya's push for Open Source

### DISCUSSION THREAD

**Discussion by codetheweb** [codetheweb/tuyapi#501] [General]
**Preface**: not sure how many of you are aware, but Tuya's recently started publishing [quite a few repos and projects](https://github.com/tuya) to GitHub. They range from OpenAPI SDK wrappers ([one of which I recently started using](https://github.com/TuyaAPI/cli/commit/a21e5d41833b88aca36f0183c59c70d162271c8e)) to C code samples for integration with their device-level SDK.

I just met with a small group of developers from Tuya, including the product manager for cloud services. A few notes:

- Their biggest concern with local control was that it turned them into a smart device factory. Their value proposition is cloud services, which projects like this nullify.
- They were even more strongly against [tuya-convert](https://github.com/ct-Open-Source/tuya-convert/tree/61ba060b8567eacd02813db431967cbe9fd23401) (see above)
- They wanted me to take on some kind of unpaid Developer Advocacy position (?) I said no.
- They're perfectly fine with individuals registering on their IoT platform to use the API and actually encourage it.
- The plan is for an official local control solution to become available around September, probably as part of their [Python API](https://github.com/tuya/tuya-iot-python-sdk) at first. **However**: it will be a binary blob. The local control component will not be open sourced. In my view this is an alright compromise but others may feel differently. They hope to make it work reliably without internet access, so you may still be able to firewall your devices off if that's your thing.

All that to be said, it didn't sound like they were planning to take legal action against this or any other project. They were impressed with the community's efforts.

They wanted ideas & feedback for their local control API since I've built one before and [started building another](https://github.com/codetheweb/tuyapi/issues/169). I figured it would be a better idea to open it up to the community, so I created a new repo that we can add ideas to: https://github.com/TuyaAPI/local-control-api-ideas/. I'll probably update it sometime tomorrow with my initial thoughts and what I'd like to see, but feedback from other community members would be fantastic as well.

**Comment by ThatWileyGuy** [comment]
I like the step overall and I'm happy to see them trying to engage with developers, but...

`Their biggest concern with local control was that it turned them into a smart device factory. Their value proposition is cloud services, 
which projects like this nullify.`

In what world do they think a round trip through US-EAST-1 is going to add value to "turning on a light" ?

Cloud services can add value to smart devices when they can orchestrate multiple devices in a way that is difficult/impossible locally or offload computation that can't be done locally on a device. As far as I'm aware, Tuya doesn't do either of those things today; the former is something they could do but don't because they make such a general framework and the latter doesn't really exist for the device types they currently support.

I'd compare most of the Tuya devices to Phillips Hue, where 99% of the applications are things like, "When I push this button I want those four lights to turn on," that are both more performant and more reliable when handled locally. The cloud service can be nice for remote access, but most of the automation I want to achieve is better when it doesn't require a round trip through a remote datacenter with all the failure points that includes.

Even for a more complicated device like an air conditioner, as long as all Tuya is achieving is off/cool/fan/dehumidify, fan speed, and temperature controls, I don't see how the cloud services add value. If Tuya were orchestrating an air conditioner running off of a separately-deployed temperature sensor, orchestrating air conditioners across rooms, or implementing a smart schedule of some sort, that would change things dramatically. To my knowledge, they aren't.

Anyway, sorry about the mini-rant, but at the end of all that, I also actively seek out Tuya devices specifically because I know I'll get reliable local control. That's a feature, not a bug. By contrast, I ran a portable AC from LG this season and I don't want to buy another one because it's so hard to control it locally.

**Comment by codetheweb** [comment]
I left a few initial thoughts: https://github.com/TuyaAPI/local-control-api-ideas.

Feel free to add others.

**Comment by gunnicom** [comment]
Ok, so if they want to be a cloud service and not a device factory. It is their decision.
Sounds to me like the "Apple" management perpective.
I do not want to buy cloud, I want to buy a device.
Conclusion: My next device won't be Tuya then.

**Reply by pkerney** [reply]
Agreed. Sounds like they want to be Apple/Amazon//Google and the millions of other IoT vendors trying to do the same thing.
I bought a "RPP06 Tuya 1080P PTZ Two Way Audio WiFi Wireless Camera" from BangGood specifically because it advertised ONVIF for NVR and browser support. You are forced to configure it with the TuyaSmart app and then the ONLY way to access is it through the app.
Then they want to sell me a cloud service for $6.99 p/m for a $35 camera. Sounds like inkjet printer ink strategy to me.
It has open ftp/telnet/8000/6668/6789 ports but the configuration gives it a password which you cannot find out.
I guess if I get local control and ONVIF stream directly, I don't pay for their service so they get nothing more from me. And that's what they will get.

**Comment by codetheweb** [comment]
Quick update, Tuya contacted me a few days ago and local control has been delayed indefinitely. They're still working on it but will not have it ready-to-go this month and don't really have any kind of deadline for it currently.

I'll keep this discussion up-to-date as I receive more information 👍.

**Comment by Dave9111** [comment]
I have several Tuya devices.   Their smart bulbs are interesting and I wanted to explore using them as a signaling device by color and on-off flashing, etc.    I went down the Tuya Convert road and found that my devices were locked.  So that was a dead end.   Then explored gathering data from the leftovers of their Smart Life app as described on Youtube.   Yep, I got the ID and key, etc but it took some effort and that key is only good for "that" installation.   Remove the bulb and start over again, I believe.   Even finding the bulbs IP address was a hassle as it didn't have a normal device info pane for some reason in Smart Life.   So I thought, hey, why not do this straight up and try their official API - it appears that some people have done that.. but not many.   So I sign up and give them my info and I get an account, start an app, and find that unless I pay them $5000 my app goes away in 30 days.    So much for open API!    Then I find out that IFTT has a Smart Life interface.    So my RPI -> IFTT -> Tuya server -> light bulb ???   Ok... maybe for now.   Nope, IFTT has what looks like a Smart Life front end, but its not functional.   Apparently discontinued, likely by Tuya.
Obviously Tuya wants a closed ecosystem where they are calling the shots.   I have wasted two days of my life trying to make their products work for me and it has been a total failure.  This is really unfortunate as their bulbs have a lot more potential than what they are selling.   What is wrong with selling smart devices that have a wide variety of uses??  I have 6 Tuya Feit bulbs in packages and their value has now greatly diminished.  I've lost all interest in Tuya.   All they have is "dumb" smart stuff that is extremely limited.   They might have cornered the market in "dumb" smart stuff, but their future is dim.    Fortunately Tuya has some competition and I bought a few test bulbs that are already flashed with an open interface that can be controlled locally.   They are on their way.  They cost more than twice what the Tuya bulbs cost.
Once my project is finished I'll opensource the project and specify the bulbs which should drive the sale of thousands of bulbs.
So long Tuya...
